tests/functional/commands/test_db_dump.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
+"""Perform functional tests for db dump."""
+
+import os
+import pickle
+
+import orion.core.cli
+from orion.core.io.database.pickleddb import PickledDB
+from orion.storage.base import setup_storage
+
+
+def execute(command, assert_code=0):
+    """Execute orion command and return returncode"""
+    returncode = orion.core.cli.main(command.split(" "))
+    assert returncode == assert_code
+
+
+def clean_dump(dump_path):
+    """Delete dumped files."""
+    for path in (dump_path, f"{dump_path}.lock"):
+        if os.path.isfile(path):
+            os.unlink(path)
+
+
+def test_dump_default(three_experiments_family_same_name_with_trials, capsys):
+    """Test dump with default arguments"""
+    assert not os.path.exists("dump.pkl")
+    try:
+        execute("db dump")
+        assert os.path.isfile("dump.pkl")
+        dumped_db = PickledDB("dump.pkl")
+        with dumped_db.locked_database(write=False) as internal_db:
+            collections = set(internal_db._db.keys())
+        assert collections == {"experiments", "algo", "trials", "benchmarks"}
+        assert len(dumped_db.read("experiments")) == 3
+        assert len(dumped_db.read("algo")) == 3
+        assert len(dumped_db.read("trials")) == 24
+        assert len(dumped_db.read("benchmarks")) == 3
+    finally:
+        clean_dump("dump.pkl")
+
+
+def test_dump_unknown_experiment(
+    three_experiments_family_same_name_with_trials, capsys
+):
+    """Test dump unknown experiment"""
+    assert not os.path.exists("dump.pkl")
+    try:
+        execute("db dump -n i-dont-exist", assert_code=1)
+        captured = capsys.readouterr()
+        assert captured.err.startswith(
+            "Error: No experiment found with query {'name': 'i-dont-exist'}. Nothing to dump."
+        )
+    finally:
+        clean_dump("dump.pkl")
```

Comment:
Should `dump.pkl` exist if the command failed?

Reply:
Yes! Output file is created as soon as the destination storage object is created in `dump_database()` (PickledDB generates the PKL file immediately), and storage object is allocated before checking if experiment exists or not. So, output file will exist in any case.

Reply:
We should try to delete it whenever possible when the command fails (as part of the execution of `db dump`, not in the test cleanup phase). It's confusing otherwise.

Reply:
Done! https://github.com/Epistimio/orion/pull/1082/commits/8cd1528cf5d450f42f8c1a235addc7f4fef2aa85